src/components/toc.js

```diff
@@ -26,39 +23,41 @@ const Toc = ({ toc }) => {
 const createIntersectionObserver = () => {
 	// gripes with intersection observer: https://twitter.com/nipafx/status/1217590444772810753
 
-	// to make sure only events after the initial observation are acted on,
-	// store the ids of visited headers
-	const visited = []
-	const update = events =>
-		events.forEach(event => {
-			const id = event.target.querySelector(`span`).id
-			if (visited.includes(id)) {
-				// `event.isIntersecting` should indicate whether the element
-				// transitioned into or out of intersection, but Firefox instead
-				// returns true if intersection ratio is > 0;
-				// instead fall back to comparing intersection ratio with threshold
-				const visible = event.intersectionRatio < INTERSECTION_THRESHOLD
-				if (visible) {
-					const position = event.boundingClientRect.top
-					updateHighlight(id, position)
-				}
-			} else visited.push(id)
-		})
-	return new IntersectionObserver(update, {
+	// registering the intersection observer triggers an event for each
+	// observed header, but none of them should be highlighted at this point;
+	// instead store the initial events' targets and only act on an event
+	// if its target was stored earlier.
+	const visited = {}
+
+	const updateHighlights = event => {
+		const focusedHeaderId = event.target.querySelector(`span`).id
+		if (!visited[focusedHeaderId]) {
+			visited[focusedHeaderId] = true
+			return
+		}
+
+		const isHeaderVisible = event.intersectionRatio < INTERSECTION_THRESHOLD
+		if (isHeaderVisible) {
+			// `event.isIntersecting` should indicate whether the element
+			// transitioned into or out of intersection, but Firefox instead
+			// returns true if intersection ratio is > 0;
+			// instead fall back to comparing intersection ratio with threshold
+			const headerPosition = event.boundingClientRect.top
+			lowlightItems()
+			highlightItems(focusedHeaderId, headerPosition)
+		}
+	}
+
+	const onObserve = events => events.forEach(updateHighlights)
+
+	return new IntersectionObserver(onObserve, {
 		// when scrolling to a header in the toc, no event is thrown if the margin
 		// is exactly the same as the header height
 		rootMargin: `-${HEADER_HEIGHT + EXTRA_MARGIN}px 0px -50%`,
 		threshold: INTERSECTION_THRESHOLD,
 	})
 }
 
-const updateHighlight = (id, position) => {
-	// remember that the ToC is duplicated (accordion and pop-out-accordion),
-	// so each ToC entry exists twice on the page; hence "itemS" (plural)
```

Comment:
Damn, I now see that I forgot to include this comment. Once again, I think this one is important. There's a single table of contents, right, and each only ever has one highlighted item. Why the plural? Why the `querySelectorAll`? These are likely questions that the code (as it is) doesn't answer, but the comment does.